registry/storage/cache/cache.go

```diff
@@ -33,3 +40,25 @@ func ValidateDescriptor(desc distribution.Descriptor) error {
 
 	return nil
 }
+
+// Register is used to register an InitFunc for
+// a BlobDescriptorCacheProvider with the given name.
+// It's meant to be called from init() function.
+func Register(name string, initFunc InitFunc) {
```

Comment:
Yeah, this is possible, but I'm trying to follow the established conventions in this codebase (see the storage drives and middleware)

Reply:
What's so sacred about the established conventions?

Reply:
Nothing in particular, but I prefer keeping code consistent across codebases: it makes navigation, comprehension, onboarding new devs, etc much easier...I got a bit religious with it in this codebase I agree

Reply:
I agree for keeping code consistent, no code is perfect and having to deal with the same suboptimal convention everywhere is better than having to deal with 5 or 6 practices of varying quality, even if the average quality is better. 

If we did want to establish a new convention, we should be deliberate about picking one and moving forward with that single practice, and we should also look towards switching the old code over to that as well.

We don't have to rip everything out all at once, but if the new convention is not compelling enough that we'd see benefit from applying it to the older code, then it may also not be compelling enough to deal with two conventions during the interim period.

Reply:
Let me get this straight:
- All existing code has been grandfathered in as "conventional."
  - Except for the S3 storage driver. For some reason, everything it does has been decreed anti-conventional.
- All new code to which any similarity can be drawn with existing code must be shaped like the existing code.
  - Except for the S3 storage driver. [New code must be _inconsistent with_ the S3 driver, just to be contrary.](https://github.com/distribution/distribution/pull/4169#discussion_r1421043575)
- Any changes to how code is shaped must be agreed upon by consensus and must then be ratified by applying it to the entire codebase.
  - Any changes must therefore go through a waterfall process before a single line of new code can be merged which goes against grandfathered conventions.
  - The determination of whether a new convention is compelling enough must be made without having experience coding in the new convention. Experience in coding the new convention cannot be gained without coding in the new convention. New conventions are therefore de facto banned. Whatever conventions were grandfathered in are our coding conventions, now and forever.
- And given which thread this discussion is being made on, anything which might possibly _afford_ code to be written in a divergent style _while still affording the conventional style_ is **also banned**

That can't be right.

I am literally just suggesting that the _brand new_ `Register` function grow a dummy result parameter so that implementations _could_ use a less-noisy syntax for registration. The following would also compile just fine!
```go
func init() {
        cache.Register("myprovider", initFunc)
}
```

I'm here to push boundaries and shake up existing conventions in order to evolve the codebase. If existing conventions are so sacred that they cannot even be iterated upon in greenfield code, I do not see a future where I remain on as a distribution maintainer.

Reply:
> I'm here to push boundaries and shake up existing conventions in order to evolve the codebase. If existing conventions are so sacred that they cannot even be iterated upon in greenfield code, I do not see a future where I remain on as a distribution maintainer.

FYI: I appreciate your feedback very much. In fact, I really enjoy working together/collaborating on this project with you. I had no idea some of my comments would cause this much discussion or maybe even angst (hopefully not!).

This codebase does need a ton of improvements all around and I need to withhold [at least some of my] hot takes (e.g. "set it on fire one lol") here as they're not productive. This discussion should probably belong elsewhere though.

FYI2: I'm not against changing the `InitFunc` - I'm sure we can find a way to make this work 😄

Reply:
> I'm here to push boundaries and shake up existing conventions in order to evolve the codebase. If existing conventions are so sacred that they cannot even be iterated upon in greenfield code, I do not see a future where I remain on as a distribution maintainer.

Please know that I am also intensely frustrated by the way this codebase does things. This project is old enough to have existed before a lot of Go best practices were solidified, and it shows.

> Any changes to how code is shaped must be agreed upon by consensus

You need two approvals from two maintainers in different orgs, in your case that's yourself and one other maintainer.

I tend to be more cautious, and I think it's a good and healthy balance to have someone who wants to shake things up. We need to moderate each other in this way. 

> and must then be ratified by applying it to the entire codebase.

I have a solution below that might give us both consensus and the capacity to effect change.

>     The determination of whether a new convention is compelling enough must be made without having experience coding in the new convention. Experience in coding the new convention cannot be gained without coding in the new convention. New conventions are therefore de facto banned. Whatever conventions were grandfathered in are our coding conventions, now and forever.

We all have worked on other projects and seen different conventions applied there and what their results were.
We can reasonably determine if something is going to work out and then decide to commit to it. 

If we want to adopt a new convention, we can start with a single PR. However, if there's no barrier to entry, then it's easy to have a bunch of unrelated incohesive practices sprinkled throughout the code. Part of being a maintainer on any project is thinking about the impact of changes in context of existing code. 

What I think would help in resolving these conflicts is adopting or implementing a style guide. This way, we can be deliberate about what our practices are, and apply them consistently. This would allow us to stabilize our practices moving forward and justify and encourage creating PRs which make old code conform. Having a style guide should provide the consistency that I want to see, while providing you with the ability to evolve the codebase.

To move forward, I propose:

- Bucking the convention of returning interfaces, right now, in this PR. 
- Creating a style guide with a single entry in support of this style, in this or a followup PR.
- Creating an issue to discuss expanding the style guide, for example, if we want to write our own or adopt some existing guide and if so, which one.